modyn/supervisor/internal/pipeline_executor/evaluation_executor.py

```diff
@@ -0,0 +1,331 @@
+from modyn.supervisor.internal.eval.handler import EvalHandler, EvalRequest
+from modyn.supervisor.internal.eval.result_writer.json_result_writer import JsonResultWriter
+from modyn.supervisor.internal.grpc.enums import IdType, MsgType, PipelineStage
+from modyn.supervisor.internal.grpc.template_msg import id_submsg, pipeline_stage_msg
+from modyn.supervisor.internal.grpc_handler import GRPCHandler
+from modyn.supervisor.internal.pipeline_executor.models import (
+    ConfigLogs,
+    PipelineLogs,
+    SingleEvaluationInfo,
+    StageLog,
+    SupervisorLogs,
+)
+from modyn.supervisor.internal.utils.evaluation_status_reporter import EvaluationStatusReporter
+from modyn.utils.utils import current_time_micros, dynamic_module_import
+from pydantic import BaseModel
+
+eval_strategy_module = dynamic_module_import("modyn.supervisor.internal.eval.strategies")
+
+logger = logging.getLogger(__name__)
+
+
+class EvalStateConfig(BaseModel):
+    pipeline_id: int
+    eval_dir: Path
+    config: ModynConfig
+    pipeline: ModynPipelineConfig
+
+
+@dataclass
+class AfterPipelineEvalContext:
+    tracking_dfs: dict[str, pd.DataFrame]
+
+
+class EvaluationExecutor:
+    def __init__(
+        self,
+        pipeline_id: int,
+        pipeline_logdir: Path,
+        config: ModynConfig,
+        pipeline: ModynPipelineConfig,
+        grpc: GRPCHandler,
+    ):
+        self.pipeline_id = pipeline_id
+        self.pipeline_logdir = pipeline_logdir
+        self.config = config
+        self.pipeline = pipeline
+        self.grpc = grpc
+        self.context: AfterPipelineEvalContext | None = None
+        self.eval_handlers = (
+            [EvalHandler(eval_handler_config) for eval_handler_config in pipeline.evaluation.handlers]
+            if pipeline.evaluation
+            else []
+        )
+
+    def register_tracking_info(self, tracking_dfs: dict[str, pd.DataFrame]) -> None:
+        assert tracking_dfs.get(PipelineStage.HANDLE_SINGLE_TRIGGER.name) is not None
+        assert tracking_dfs.get(PipelineStage.STORE_TRAINED_MODEL.name) is not None
+        self.context = AfterPipelineEvalContext(tracking_dfs=tracking_dfs)
+
+    def create_snapshot(self) -> None:
+        """Create a snapshot of the evaluation state before starting to evaluate."""
+        if not self.pipeline.evaluation:
+            return
+
+        snapshot_dir = self.pipeline_logdir / "snapshot"
+
+        # create tempdir if snapshot_dir is None:
+        snapshot_dir.mkdir(parents=True, exist_ok=True)
+
+        # write state: config, pipeline & context
+        eval_state_config = EvalStateConfig(
+            pipeline_id=self.pipeline_id, eval_dir=self.pipeline_logdir, config=self.config, pipeline=self.pipeline
+        )
+        (snapshot_dir / "eval_state.yaml").write_text(eval_state_config.model_dump_json(by_alias=True))
+        (snapshot_dir / "context.pcl").write_bytes(pickle.dumps(self.context))
+
```

Comment:
If my understanding is correct that this is pipeline metadata, can't we write this into the pipeline logfile and then just restart the evaluation using the pipeline logfile? I'd try avoiding creating many files per pipeline and rather have one condensed summary of the pipeline